### ISSUE TITLE
Index Hierarchy set operations lose index class information

### ISSUE THREAD

**Issue by chaburkland** [InvestmentSystems/static-frame#217]
This is pretty minor issue, but does have some unexpected lossy behavior. When you perform index set operations (`union`, `intersection`, `difference`) between two indices and a level was built with `IndexDate`, the result does not retain that class information.

See below, where the inner level was originally `IndexDate` between the two indices, but after the operation, it is changed to the default `Index`.

```python3
>>> i1 = sf.IndexHierarchy.from_labels([[1, date(2019, 1, 1)], [2, date(2019, 1, 2)]], index_constructors=[sf.Index, sf.IndexDate])
>>> i2 = sf.IndexHierarchy.from_labels([[2, date(2019, 1, 2)], [3, date(2019, 1, 3)]], index_constructors=[sf.Index, sf.IndexDate])
>>> i1
<IndexHierarchy>
1                2019-01-01
2                2019-01-02
<int64>          <datetime64[D]>

>>> i2
<IndexHierarchy>
2                2019-01-02
3                2019-01-03
<int64>          <datetime64[D]>

>>> i1.union(i2)
<IndexHierarchy>
1                2019-01-01
2                2019-01-02
3                2019-01-03
<int64>          <object>

>>> i1.intersection(i2)
<IndexHierarchy>
2                2019-01-02
<int64>          <object>

>>> i1.difference(i2)
<IndexHierarchy>
1                2019-01-01
<int64>          <object>
```

Note: This issue only applies to IndexHierarchies. Index class is retained with normal index operations.

**Comment by flexatone** [comment]
Many thanks for isolating this issue!

I will explore a fix.

**Comment by flexatone** [comment]
Thanks again for this issue.

``IndexHierarchy`` set operations will now delegate ``Index`` types where they are equivalent between operands at corresponding depth levels. While restrictive, this seems like the only sensible approach to take if the ``Index`` types are different. This change will be included in the 0.6.21 release.

```python
>>> from datetime import date
>>> i1 = sf.IndexHierarchy.from_labels([[1, date(2019, 1, 1)], [2, date(2019, 1, 2)]], index_constructors=[sf.Index, sf.IndexDate])
>>> i2 = sf.IndexHierarchy.from_labels([[2, date(2019, 1, 2)], [3, date(2019, 1, 3)]], index_constructors=[sf.Index, sf.IndexDate])
>>> i3 = i1.union(i2)
>>> i3
<IndexHierarchy>
1                2019-01-01
2                2019-01-02
3                2019-01-03
<int64>          <datetime64[D]>
>>> i3.index_types
<Series>
<Index>
0        <Index>
1        <IndexDate>
<int64>  <object>
```
